Bring up-to-date with base (fixes #13)

RyanGlScott opened the pull request:
This applies a number of changes to make `generic-deriving` consistent with `base`, namely:

* Derives `Eq`, `Ord`, `Read`, and `Show` instances for generics-related data types (added to `base` in 4.7.0.0).
* Makes the backported `Generic`/`Generic1` instances consistent with what `base` has (or, if they're derived, I made them consistent with what `-ddump-deriv` displays).

RyanGlScott commented:
Note that I only backported `Generic` instances back to GHC < 7.1 if they were already present in `generic-deriving` (i.e., the instances for `Char`, `Int`, `Float`, `Maybe`, and `[]`), even though they are more instances than that. If you want me to backport all of those instances, let me know.

dreixel commented:
Thanks for this. But I'm wondering: does `generic-deriving` itself require `base-orphans` to compile? If not, why introduce the dependency, instead of letting users decide by themselves whether they want to depend on `base-orphans` or not?

RyanGlScott commented:
The reason I use `base-orphans` is to prevent potential conflicts with orphan instances on GHC 7.2 and 7.4. Previously, `generic-deriving` defined orphan `Generic1` instances for `Maybe` and `[]` using `Generic1` from `ghc-prim`, which means that if anybody else were to also define these instances in their own packages, it could potentially lead to instance conflicts down the road if they used their library and `generic-deriving` together. As an example of this, I recently used `base-orphans` to consolidate orphan instances for `Foldable ((,) a` between the [`functor-combo`](https://github.com/conal/functor-combo/pull/2) and [`infernu`](https://github.com/sinelaw/infernu/pull/18) pacakges (before, you simply couldn't use them together on GHC 7.6 and earlier because of the duplicate `Foldable ((,) a` orphan instances).

Alternatively, you could define orphan `Generic1 Maybe` and `Generic1 []` instances in `generic-deriving` on GHC 7.2/7.4. That means anyone wanting to have backwards-compatible `Generic1 Maybe` and `Generic1 []` instances in their library would have to depend on `generic-deriving` to avoid instance conflicts. `generic-deriving`, however, has more dependencies than `base-orphans`, so I feel that `base-orphans` is a better place to put these instances.

RyanGlScott commented:
It's up to you whether you'd want it as a dependency or not. If you don't, let me know—currently, `generic-deriving` can't be used with any packages that have `base-orphans` as a dependency for the above reasons, so I'd have to come up with a workaround.

RyanGlScott commented:
Now that `semigroupoids` has `base-orphans` as a dependency, there's at least one library on Hackage with both `generic-deriving` and `base-orphans` as transitive dependencies which fails to build on GHC 7.4: [`log-domain`](http://hackage.haskell.org/package/log-domain-0.10.1):

```
[1 of 2] Compiling Numeric.Log      ( src/Numeric/Log.hs, dist/dist-sandbox-d6d975f1/build/Numeric/Log.o )

src/Numeric/Log.hs:1:1:
    Conflicting family instance declarations:
      type Rep1 [] -- Defined in `base-orphans-0.3.2:Data.Orphans'
      type Rep1 [] -- Defined in `Generics.Deriving.Instances'

src/Numeric/Log.hs:1:1:
    Conflicting family instance declarations:
      type Rep1 Maybe -- Defined in `base-orphans-0.3.2:Data.Orphans'
      type Rep1 Maybe -- Defined in `Generics.Deriving.Instances
```

dreixel commented:
What would be the workaround?

RyanGlScott commented:
The root of the problem is `generic-deriving` and `base-orphans` both backporting the same `Generic` and `Generic1` instances on GHC 7.2 and 7.4, so I can think of three different solutions:

1. Make `generic-deriving` depend on `base-orphans` and remove the backported instances from `generic-deriving` (this is what this pull request accomplishes). I'd recommend this approach since [other libraries using `base-orphans`](http://packdeps.haskellers.com/reverse/base-orphans) wouldn't have to change.
2. Don't depend on `base-orphans`. Instead, remove the backported `Generic`/`Generic1` instances from `base-orphans`. This "works", except that it would necessarily involve a breaking change to `base-orphans`, and users would have to depend on `generic-deriving` in order to bring certain backported `Generic`/`Generic1` instances into scope, which feels off to me.
3. Don't depend on `base-orphans`. Instead, remove the backported `Generic`/`Generic1` instances from `generic-deriving`. This fixes the problem at the cost of conditionally exporting instances depending on the version of GHC you're using, which also feels off to me.

I'd vouch for approach 1, but again, it's up to you.

RyanGlScott commented:
After thinking about it some more, it's probably best if `generic-deriving` were the canonical source of these orphan instances. This way, `generic-deriving` will continue to support GHCs prior to 7.0 (which `base-orphans` does not), and `generic-deriving` won't need an extra dependency.

I've updated the pull request to backport all `Generic` and `Generic1` instances introduced in more recent versions of `base`, and I removed the `base-orphans` dependency. @dreixel, can you review and make sure it looks good?

dreixel commented:
I can't say I have looked into the details of every instance; I trust you got them right. Thanks for all the work and for pressing to get this fixed!

RyanGlScott commented:
Thanks, Pedro!

All of those instances (with the exception of `Char`, `Int`, `Double` and `Float`, which were copied from `base`), were copied from the output of `-ddump-deriv` when compiling standalone `deriving instance Generic ...` statements on GHC 7.10.1, so they should be fine.
